src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -560,6 +560,14 @@ private BoundExpression BindSimpleBinaryOperator(BinaryExpressionSyntax node, Di
 
             if (hasErrors)
             {
+                if (GetTupleCardinality(left).HasValue &&
+                    GetTupleCardinality(right).HasValue &&
+                    (kind == BinaryOperatorKind.Equal || kind == BinaryOperatorKind.NotEqual))
+                {
+                    CheckFeatureAvailability(node, MessageID.IDS_FeatureTupleEquality, diagnostics);
+                    return BindTupleBinaryOperator(node, kind, left, right, diagnostics);
```

Comment:
>return BindTupleBinaryOperator(node, kind, left, right, diagnostics) [](start = 20, length = 68)

It feels wrong to call this when we already detected that there is an error and possibly ignore the fact. Also it looks like we might have done a lot of things above (created some conversions, etc.) that already could have produced errors that won't be expected if we are able to successfully bind tuple equality/inequality. I think that tuple equality should be handled similar  to other built-in operators and should be folded into logic in BinaryOperatorOverloadResolution.

Reply:
This also feels very fragile considering backward compatibility maintenance in the future. Future changes in BinaryOperatorOverloadResolution could easily make scenarios that fail resolution today and, therefore, treated as tuple equality today to become success cases and no longer treated as tuple equality.

---
In reply to: [170487145](https://github.com/dotnet/roslyn/pull/23356#discussion_r170487145) [](ancestors = 170487145)

Reply:
Thanks. There is likely a bug with diagnostics from conversions.
As for the overall flow, this matches current spec'ed and discussed behavior. The C# spec lists out 3 cases: (1) dynamic, (2) binary overload resolution, and (3) error. The proposal was to add a new case between (2) and (3).
That said, I need to think more about your idea. I'll add a PROTOTYPE marker and investigate this week.

---
In reply to: [170488759](https://github.com/dotnet/roslyn/pull/23356#discussion_r170488759) [](ancestors = 170488759,170487145)